### DISCUSSION TITLE
Symmetry error - why?

### DISCUSSION THREAD

**Discussion by tracylynne99** [jqno/equalsverifier#519] [Q&A] [answered]
So I'm testing the use of EqualsVerifier against a jfreechart class: CategoryItemEntity. Fixed a couple of errors found by EqualsVerifier, but now I'm getting this output and I'm stuck on how to fix it.

> Testcase: equalsContract(org.jfree.chart.entity.CategoryItemEntityTest):	FAILED
Symmetry:
  CategoryItemEntity: rowKey=nl.jqno.equalsverifier.internal.reflection.Comparable$$DynamicSubclass$32917505@1f1cf9, columnKey=nl.jqno.equalsverifier.internal.reflection.Comparable$$DynamicSubclass$32917505@1f1cf9, dataset=org.jfree.data.category.CategoryDataset$$DynamicSubclass$3590374@3f6c10
does not equal superclass instance
  ChartEntity: tooltip = one
For more information, go to: http://www.jqno.nl/equalsverifier/errormessages
junit.framework.AssertionFailedError: Symmetry:
  CategoryItemEntity: rowKey=nl.jqno.equalsverifier.internal.reflection.Comparable$$DynamicSubclass$32917505@1f1cf9, columnKey=nl.jqno.equalsverifier.internal.reflection.Comparable$$DynamicSubclass$32917505@1f1cf9, dataset=org.jfree.data.category.CategoryDataset$$DynamicSubclass$3590374@3f6c10
does not equal superclass instance
  ChartEntity: tooltip = one
	at nl.jqno.equalsverifier.EqualsVerifier.handleError(EqualsVerifier.java:375)
	at nl.jqno.equalsverifier.EqualsVerifier.verify(EqualsVerifier.java:361)
	at org.jfree.chart.entity.CategoryItemEntityTest.equalsContract(CategoryItemEntityTest.java:67)

I did read the explanation at the url, but I'm still lost.

Here's the JUnit test (I tried copying/pasting the code but it looked terrible, even when I surrounded it by the markup tic, so here's some screenshots):
![image](https://user-images.githubusercontent.com/4762080/137361099-5ddfba67-b018-4994-ae30-1ebb380748c2.png)

Here's the hashCode method for CategoryItemEntity:
![image](https://user-images.githubusercontent.com/4762080/137361368-f38e5c62-9181-4a71-919a-7edcaf08f8fc.png)

And the equals method: 
![image](https://user-images.githubusercontent.com/4762080/137361510-e78b55b1-66c1-41b9-8616-e99fec50f9bd.png)

When I ran the JUnit test in debug mode, I set a breakpoint at line 230 in CategoryItemEntity so I could see what the values of "rowKey" were.  One was null, the other was this nl.jqno.equalsverifier.internal.reflection.Comparable$$DynamicSubclass.

Any help would be appreciated!

Thanks!

**Comment by jqno** [comment]
Hi!
From the info you give, unfortunately I can't tell either. I would need to know more about the types of the fields that are being compared. Can you maybe just post the full code of the class you're testing? That would make it sooooo much easier for me to debug this issue. Doesn't matter if it isn't formatted properly, I can run a formatter, but please do include any import statements and dependencies. (Or take out the dependencies, but make sure the error is still there :) )

**Reply by tracylynne99** [reply]
Ok, here's the JUnit Test:

`/* ===========================================================
 * JFreeChart : a free chart library for the Java(tm) platform
 * ===========================================================
 *
 * (C) Copyright 2000-2016, by Object Refinery Limited and Contributors.
 *
 * Project Info:  http://www.jfree.org/jfreechart/index.html
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * [Oracle and Java are registered trademarks of Oracle and/or its affiliates. 
 * Other names may be trademarks of their respective owners.]
 *
 * ---------------------------
 * CategoryItemEntityTest.java
 * ---------------------------
 * (C) Copyright 2004-2016, by Object Refinery Limited and Contributors.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 * Contributor(s):   -;
 *
 * Changes
 * -------
 * 20-May-2004 : Version 1 (DG);
 *
 */

package org.jfree.chart.entity;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.awt.geom.Rectangle2D;
import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;

import org.jfree.chart.TestUtils;

import org.jfree.data.category.DefaultCategoryDataset;
import org.junit.Test;

/**
 * Tests for the {@link CategoryItemEntity} class.
 */
public class CategoryItemEntityTest {

    @Test
    public void equalsContract()
    {
        EqualsVerifier.forClass(ChartEntity.class)
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify(); // superclass
        EqualsVerifier.forClass(CategoryItemEntity.class)
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify();
    }

    /**
     * Confirm that the equals method can distinguish all the required fields.
     */
    @Test
    public void testEquals() {
        DefaultCategoryDataset d = new DefaultCategoryDataset();
        d.addValue(1.0, "R1", "C1");
        d.addValue(2.0, "R1", "C2");
        d.addValue(3.0, "R2", "C1");
        d.addValue(4.0, "R2", "C2");
        CategoryItemEntity e1 = new CategoryItemEntity(new Rectangle2D.Double(
                1.0, 2.0, 3.0, 4.0), "ToolTip", "URL", d, "R2", "C2");
        CategoryItemEntity e2 = new CategoryItemEntity(new Rectangle2D.Double(
                1.0, 2.0, 3.0, 4.0), "ToolTip", "URL", d, "R2", "C2");
        assertTrue(e1.equals(e2));

        e1.setArea(new Rectangle2D.Double(4.0, 3.0, 2.0, 1.0));
        assertFalse(e1.equals(e2));
        e2.setArea(new Rectangle2D.Double(4.0, 3.0, 2.0, 1.0));
        assertTrue(e1.equals(e2));

        e1.setToolTipText("New ToolTip");
        assertFalse(e1.equals(e2));
        e2.setToolTipText("New ToolTip");
        assertTrue(e1.equals(e2));

        e1.setURLText("New URL");
        assertFalse(e1.equals(e2));
        e2.setURLText("New URL");
        assertTrue(e1.equals(e2));
    }

    /**
     * Confirm that cloning works.
     */
    @Test
    public void testCloning() throws CloneNotSupportedException {
        DefaultCategoryDataset d = new DefaultCategoryDataset();
        d.addValue(1.0, "R1", "C1");
        d.addValue(2.0, "R1", "C2");
        d.addValue(3.0, "R2", "C1");
        d.addValue(4.0, "R2", "C2");
        CategoryItemEntity e1 = new CategoryItemEntity(new Rectangle2D.Double(
                1.0, 2.0, 3.0, 4.0), "ToolTip", "URL", d, "R2", "C2");
        CategoryItemEntity e2 = (CategoryItemEntity) e1.clone();
        assertTrue(e1 != e2);
        assertTrue(e1.getClass() == e2.getClass());
        assertTrue(e1.equals(e2));
    }

    /**
     * Serialize an instance, restore it, and check for equality.
     */
    @Test
    public void testSerialization() {
        DefaultCategoryDataset d = new DefaultCategoryDataset();
        d.addValue(1.0, "R1", "C1");
        d.addValue(2.0, "R1", "C2");
        d.addValue(3.0, "R2", "C1");
        d.addValue(4.0, "R2", "C2");
        CategoryItemEntity e1 = new CategoryItemEntity(new Rectangle2D.Double(
                1.0, 2.0, 3.0, 4.0), "ToolTip", "URL", d, "R2", "C2");
        CategoryItemEntity e2 = (CategoryItemEntity) TestUtils.serialised(e1);
        assertEquals(e1, e2);
    }

}
`

And this is the class:
`/* ===========================================================
 * JFreeChart : a free chart library for the Java(tm) platform
 * ===========================================================
 *
 * (C) Copyright 2000-2021, by Object Refinery Limited and Contributors.
 *
 * Project Info:  http://www.jfree.org/jfreechart/index.html
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * [Oracle and Java are registered trademarks of Oracle and/or its affiliates. 
 * Other names may be trademarks of their respective owners.]
 *
 * -----------------------
 * CategoryItemEntity.java
 * -----------------------
 * (C) Copyright 2002-2021, by Object Refinery Limited and Contributors.
 *
 * Original Author:  David Gilbert (for Object Refinery Limited);
 * Contributor(s):   Richard Atkinson;
 *                   Christian W. Zuckschwerdt;
 *
 */

package org.jfree.chart.entity;

import java.awt.Shape;
import java.io.Serializable;
import java.util.Objects;
import org.jfree.chart.util.Args;

import org.jfree.data.category.CategoryDataset;

/**
 * A chart entity that represents one item within a category plot.
 */
public class CategoryItemEntity extends ChartEntity
        implements Cloneable, Serializable {

    /** For serialization. */
    private static final long serialVersionUID = -8657249457902337349L;

    /** The dataset. */
    private CategoryDataset dataset;

    /**
     * The row key.
     */
    private Comparable rowKey;

    /**
     * The column key.
     */
    private Comparable columnKey;

    /**
     * Creates a new entity instance for an item in the specified dataset.
     *
     * @param area  the 'hotspot' area ({@code null} not permitted).
     * @param toolTipText  the tool tip text.
     * @param urlText  the URL text.
     * @param dataset  the dataset ({@code null} not permitted).
     * @param rowKey  the row key ({@code null} not permitted).
     * @param columnKey  the column key ({@code null} not permitted).
     */
    public CategoryItemEntity(Shape area, String toolTipText, String urlText,
            CategoryDataset dataset, Comparable rowKey, Comparable columnKey) {
        super(area, toolTipText, urlText);
        Args.nullNotPermitted(dataset, "dataset");
        this.dataset = dataset;
        this.rowKey = rowKey;
        this.columnKey = columnKey;
    }

    /**
     * Returns the dataset this entity refers to.  This can be used to
     * differentiate between items in a chart that displays more than one
     * dataset.
     *
     * @return The dataset (never {@code null}).
     *
     * @see #setDataset(CategoryDataset)
     */
    public CategoryDataset getDataset() {
        return this.dataset;
    }

    /**
     * Sets the dataset this entity refers to.
     *
     * @param dataset  the dataset ({@code null} not permitted).
     *
     * @see #getDataset()
     */
    public void setDataset(CategoryDataset dataset) {
        Args.nullNotPermitted(dataset, "dataset");
        this.dataset = dataset;
    }

    /**
     * Returns the row key.
     *
     * @return The row key (never {@code null}).
     *
     * @see #setRowKey(Comparable)
     */
    public Comparable getRowKey() {
        return this.rowKey;
    }

    /**
     * Sets the row key.
     *
     * @param rowKey  the row key ({@code null} not permitted).
     *
     * @see #getRowKey()
     */
    public void setRowKey(Comparable rowKey) {
        this.rowKey = rowKey;
    }

    /**
     * Returns the column key.
     *
     * @return The column key (never {@code null}).
     *
     * @see #setColumnKey(Comparable)
     */
    public Comparable getColumnKey() {
        return this.columnKey;
    }

    /**
     * Sets the column key.
     *
     * @param columnKey  the column key ({@code null} not permitted).
     *
     * @see #getColumnKey()
     */
    public void setColumnKey(Comparable columnKey) {
        this.columnKey = columnKey;
    }

    /**
     * Returns a string representing this object (useful for debugging
     * purposes).
     *
     * @return A string (never {@code null}).
     */
    @Override
    public String toString() {
        return "CategoryItemEntity: rowKey=" + this.rowKey
               + ", columnKey=" + this.columnKey + ", dataset=" + this.dataset;
    }

//    /**
//     * Tests the entity for equality with an arbitrary object.
//     *
//     * @param obj  the object ({@code null} permitted).
//     *
//     * @return A boolean.
//     */
//    @Override
//    public boolean equals(Object obj) {
//        if (obj == this) {
//            return true;
//        }
//        if (!(obj instanceof CategoryItemEntity)) {
//            return false;
//        }
//        CategoryItemEntity that = (CategoryItemEntity) obj;
//        if (!this.rowKey.equals(that.rowKey)) {
//            return false;
//        }
//        if (!this.columnKey.equals(that.columnKey)) {
//            return false;
//        }
//        if (!Objects.equals(this.dataset, that.dataset)) {
//            return false;
//        }
//
//        return super.equals(obj);
//    }

    @Override
    public int hashCode()
    {
        int hash = super.hashCode();
        hash = 37 * hash + Objects.hashCode(this.dataset);
        hash = 37 * hash + Objects.hashCode(this.rowKey);
        hash = 37 * hash + Objects.hashCode(this.columnKey);
        return hash;
    }

    @Override
    public boolean equals(Object obj)
    {
        if (this == obj)
        {
            return true;
        }
        if (obj == null)
        {
            return false;
        }
        if (!(obj instanceof CategoryItemEntity))
        {
            return false;
        }
        final CategoryItemEntity other = (CategoryItemEntity) obj;
        if (!Objects.equals(this.dataset, other.dataset))
        {
            return false;
        }
        if (!Objects.equals(this.rowKey, other.rowKey))
        {
            return false;
        }
        if (!Objects.equals(this.columnKey, other.columnKey))
        {
            return false;
        }
        return true;
    }

}
`

**Comment by tracylynne99** [comment]
I did do some RTFM since then, and now I'm wondering if it could be something about the superclass.  It's JFreeChart 1.5.3, so if you want to download the source for that release it's here: https://github.com/jfree/jfreechart/releases/tag/v1.5.3

Thanks for your help!

**Comment by jqno** [comment, marked as answer]
I see it now.

`CategoryItemEntity` implements `equals`. But it overrides `ChartEntity` which also implements `equals`. That means that we're talking about a type that adds state. That's the hardest thing to get right with EqualsVerifier, unfortunately. I have a short bit about that [on the website](https://jqno.nl/equalsverifier/manual/inheritance/), but you've probably already read that. I have a [Devoxx talk](https://www.youtube.com/watch?v=pNJ_O10XaoM) that goes into a little more depth, if you prefer video.

Basically, if you don't want to change the existing implementation of `equals`, you need to add `.withRedefinedSuperclass()` on the `CategoryItemEntity` test, and `.withRedefinedSubclass(CategoryItemEntity.class)` on the `ChartEntity` test.

**Reply by tracylynne99** [reply]
Right! I was actually playing around with this yesterday using the example classes in that article that you referenced. Got the EqualsVerifier to finally be ok with them.

I really do like this library though! For fun, I wrote a bunch of assert statements in a JUnit test (for the example classes) - one for each different item in the equals contract. Ran JUnit tests with JaCoCoverage. Got pretty decent coverage results. Then I added just a few lines using EqualsVerifier, commented out the original JUnit assert statements and reran with the coverage tool.  The coverage was even better! I'm sure you knew that already lol but it's really neat to see that!

Anyway, thanks again for your help - sorry if I wasted your time (insert face palm emoji here) :-)

**Comment by jqno** [comment]
Indeed, I make a point of it that EqualsVerifier should give 100% test coverage for equals methods :smile: 

Glad I could help!
